### PR TITLE
[ntuple] reduce memory allocations in BindRawPtr

### PR DESCRIPTION
**Pull request by jblomer**
Adds a dummy shared pointer to RValue: the dummy shared pointer holds the control block with a noop deleter if BindRawPtr is used. The object shared pointer, in this case, shares the control block with the dummy shared pointer (aliasing constructor). In this way, we avoid a memory allocation for the control block every time BindRawPtr is called.

Also some minor improvements to the internals of REntry.

@Dr15Jones FYI

### CODE REVIEW

File: tree/ntuple/v7/inc/ROOT/RField.hxx [hunk 1]
```diff
@@ -199,6 +202,12 @@ public:
       void Read(NTupleSize_t globalIndex) { fField->Read(globalIndex, fObjPtr.get()); }
       void Read(RClusterIndex clusterIndex) { fField->Read(clusterIndex, fObjPtr.get()); }
       void Bind(std::shared_ptr<void> objPtr) { fObjPtr = objPtr; }
+      void BindRawPtr(void *rawPtr)
+      {
+         if (!fRawPtrCtrlBlock)
+            fRawPtrCtrlBlock = std::shared_ptr<void>(this, [](void *) {});
```
Comment (hahnjo) [inline tree/ntuple/v7/inc/ROOT/RField.hxx:208]: Do we need this? In my tests, it was fine to alias an empty `std::shared_pointer`...

Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/RField.hxx:208]: That's probably right. It is a bit confusing that in this way we create non-empty shared pointer with `use_count() == 0`. Perhaps that's exactly what it is though.

File: tree/ntuple/v7/inc/ROOT/REntry.hxx [hunk 2]
```diff
@@ -48,6 +48,17 @@ class REntry {
...
+   /// The field token identifies a top-level field in this entry. It can be used for fast indexing in REntry's
+   /// methods, e.g. BindValue
+   class RFieldToken {
+      friend class REntry;
+      std::size_t fIndex;     ///< the index in fValues that belongs to the top-level field
+      std::uint64_t fModelId; ///< Safety check to prevent tokens from other models being used
+      RFieldToken(std::size_t index, std::uint64_t modelId) : fIndex(index), fModelId(modelId) {}
+   };
+
+private:
```
[the remark this answers is not in the document]
Comment (jblomer) [reply tree/ntuple/v7/inc/ROOT/REntry.hxx]: I'll open a separate PR.

**Comment by hahnjo** [comment]
Changes LGTM, but Windows isn't happy. I suppose this is because of the `static` variable... @bellenot can you help here?

**Comment by bellenot** [comment]
> Changes LGTM, but Windows isn't happy. I suppose this is because of the `static` variable... @bellenot can you help here?

I'll take a look

**Comment by jblomer** [comment]
Thank you! I made the modification. Still not clear to me, why it is necessary. But this is not a (very) performance-critical path, so it's fine.

**Review by hahnjo — APPROVED** [review]

**Comment by hahnjo** [comment]
(minor nit on the commit message: in the latest version, we don't allocate a control block anymore at all; the `fObjPtr` aliases an empty `shared_ptr`)
